core/trino-main/src/main/java/io/trino/operator/exchange/PartitioningExchanger.java

```diff
@@ -75,18 +86,24 @@ private synchronized void partitionPage(Page page, Page partitionPage)
             if (partitionSize == 0) {
                 continue;
             }
-            Page pageSplit;
+            // reset the assigned positions list size for the next iteration to start empty
+            positions.clear();
             if (partitionSize == page.getPositionCount()) {
-                pageSplit = page; // entire page will be sent to this partition, no copies necessary
-            }
-            else {
-                pageSplit = page.copyPositions(positions.elements(), 0, partitionSize);
+                // entire page will be sent to this partition, release the lock before compaction
+                return buffers.get(partition);
             }
-            // clear the assigned positions list for the next iteration to start empty
-            positions.clear();
-            memoryManager.updateMemoryUsage(pageSplit.getRetainedSizeInBytes());
-            buffers.get(partition).accept(new PageReference(pageSplit, 1, onPageReleased));
+            Page pageSplit = page.copyPositions(positions.elements(), 0, partitionSize);
```

Comment:
Copying is as expensive as compaction. I think we should:
1. either perform positions copy outside of lock too
or
2. Just do compaction here.

I think 2) is better and then potentially fix it in subsequent PR

Reply:
Unfortunately `copyPositions` _can't_ be performed outside of the lock because the lock is protecting the `IntArrayList[] positionAssignments` (and the inner `IntArrayList` entries) from another thread entering the partitioning loop and overwriting the position values before it has finished using them.

The single partition case is special in that we know once we've found that situation that we're done with the positions lists and it's safe for other threads to proceed. That said, compacting the page inside of the synchronized block is still better compared to not, so I'm ok with leaving the compaction inside the synchronized region for now (although I would still plan to change it back afterwards). Let me know if you'd still prefer to fix the issue with approach 2 for now.

Reply:
> Unfortunately copyPositions can't be performed outside of the lock because the lock is protecting the IntArrayList[] positionAssignments (and the inner IntArrayList entries) from another thread entering the partitioning loop and overwriting the position values before it has finished using them.

Is it really faster to reuse that arrays?

> That said, compacting the page inside of the synchronized block is still better compared to not, so I'm ok with leaving the compaction inside the synchronized region for now (although I would still plan to change it back afterwards). Let me know if you'd still prefer to fix the issue with approach 2 for now.

Let's split it into two PRs

Reply:
> Let's split it into two PRs
Sure, will do.

> Is it really faster to reuse that arrays?
I think this is more of an allocation rate and total footprint consideration. If each driver feeding the `LocalExchangeSink` had it's own partitioning arrays that would be `task_concurrency` * `task_concurrency` instances of `IntArrayList`, each retaining an `int[]` the size of the largest input page position counts seen. By sharing the `IntArrayList[] partitionAssignments` that's a one-time scratch structure with access guarded by a mutex. I think there should be a way to do the initial page partition assignments outside of the critical section to reduce the duration of the lock being held, but that's a more dramatic experimental change to make.